### PR TITLE
custom.py: fix #7512 luajit<2.1 issue on Catalina

### LINKED ISSUE #7512: [master branch] Could not initialize Lua when luajit < 2.1.0 installed on macOS Catalina
### mpv version and platform versions
`master` branch, Mac OS X 10.15.33 

### Reproduction steps

this behavior happens after https://github.com/mpv-player/mpv/commit/d852ad2ae76c97fce710e41064c863842db87035 

and only happens when both `luajit` and `lua@5.1` are installed from Homebrew
```bash
brew install luajit
brew install lua@5.1
```

### Expected behavior

in my `mpv.conf`, I added one line as 
```
script-opts=osc-layout=box
```

from tag `release/0.32.0` or any commit before https://github.com/mpv-player/mpv/commit/d852ad2ae76c97fce710e41064c863842db87035 

I can get the OSC show up when I move my cursor

Screen Shot 2020-03-07 at 08 30 21
 

### Actual behavior

But when I build from https://github.com/mpv-player/mpv/commit/d852ad2ae76c97fce710e41064c863842db87035  or any commit after this commit on master branch. The OSC could never show up

Screen Shot 2020-03-07 at 08 31 19

### Log file

https://0x0.st/iT9G.txt

The issue will be closed for ignoring the issue template.

### My thought

This issue ONLY happens when BOTH `luajit` and `lua@5.1` are installed from Homebrew.

Before https://github.com/mpv-player/mpv/commit/d852ad2ae76c97fce710e41064c863842db87035 , the build script will pick up `lua@5.1` first, after this commit, it will pick up `luajit`

I try to uninstall `luajit` from my local and build from `master` again,  the OSC comes back.

I try to revert this commit on my local and build it again with both `luajit` and `lua@5.1` installed,  the OSC works well.

I am not sure whether it is a compatible issue or a configuration issue.

A quick fix is revert this commit https://github.com/mpv-player/mpv/commit/d852ad2ae76c97fce710e41064c863842db87035 , or at least move `lua@5.1` on the top of `luajit` . I can create this PR if necessary. 

A long term solution seems like require some time to investigate further.

cc @wm4

### PR DESCRIPTION
**Pull request by jnozsc**
fix #7512 

I have verified this on my local, this change works

### CODE REVIEW

**Comment by Akemi** [comment]
i personally would have liked the changes for the actual fix separated from the changes for python compatibility and style changes. otherwise i am not sure this is the right approach to handle this problem.

also the travis build broke. though no idea with that error message. maybe an hiccup again.

**Comment by jnozsc** [comment]
thanks for review, I revert the style change in https://github.com/mpv-player/mpv/pull/7574/commits/8ecfed37954b429ba3eae82065e297e5502ed64b

**Comment by jnozsc** [comment]
2 xcode tests complain as `An error occurred while generating the build script.` is there any way to trigger them to run again without create a new commit?

Thanks

**Comment by jnozsc** [comment]
I modify the travis in order to ensure luajit >= 2.1 is installed on mac SDK >= 10.15 (in travis, the image name is "xcode11.3" https://docs.travis-ci.com/user/reference/osx/#xcode-1131 )

**Comment by Akemi** [comment]
tbh, this should be rejected. first it's completely not our fault. it's either a broken luajit build or version, and that only for macOS 10.15. it would be ridiculous to add such a check to every project downstream that uses luajit. furthermore luajit didn't tag a new release yet and building from HEAD in the CI should be discouraged (the macOS build parts are already quite complicated).

this should be fixed upstream. either in the package manager (which distributes a broke build) or in the project itself.

**Comment by jnozsc** [comment]
`luajit` is totally silent https://github.com/LuaJIT/LuaJIT/issues/563
`homebrew` does not want to fix it https://github.com/Homebrew/homebrew-core/pull/51335

Before I create this PR, I have already reached out these upstream you mentioned, but it seems they are not interested on fixing it

**Comment by avih** [comment]
> it seems they are not interested on fixing it

Then why should we be interested? The issue is clearly not at the mpv side, and "fixing" it at the mpv side is inappropriate IMO.

You can add something to the mpv wiki pages that on Catalina, if the system has both luajit and lua5.2 then they should configure mpv with `--lua=lua52` or some such (check the specifics before you update the wiki).

File: waftools/checks/custom.py [hunk 1]
```diff
@@ -67,13 +68,20 @@ def check_lua(ctx, dependency_identifier):
...
 
     if ctx.options.LUA_VER:
-        lua_versions = \
-            [lv for lv in lua_versions if lv[0] == ctx.options.LUA_VER]
+        lua_versions = OrderedDict(
+            [(k, v) for k, v in lua_versions.items() if k == ctx.options.LUA_VER]
+        )
 
-    for lua_version, pkgconfig_query in lua_versions:
+    # luajit < 2.1.0 is broken on macOS Catalina, issues #7512
```
Comment (avih) [inline waftools/checks/custom.py:78]: We should never do this. If it's broken on Catalina, then users should not install it on Catalina, or the package manager should disable it on Catalina.

**Comment by wm4** [comment]
LuaJIT is probably not responding because none of its maintainers (if there are any left) actually look at the github issue tracker. Maybe you would have more success posting on the luajit ML, but I don't know.

I don't think it's mpv's responsibility to work this around. If anything, this would be the job of the distro/package manager, i.e. homebrew.

**Comment by Argon-** [comment]
This is an unfortunate situation. Best solution would be a new release from LuaJIT but no such thing should be expected imo, considering it's basically unmaintained. 
Now, in such a situation it would be the responsibility of a package manager to fix this. However, homebrew doesn't consider HEAD builds as first world citizens. What you need is a package manager that does and then those people are who you need to go to.
I realize no such thing exists for macOS, that's why it's an unfortunate situation.

**Comment by avih** [comment]
Why doesn't brew make mpv depend on lua52 and just use `--lua=lua52` at its config parameters? This would solve all issues without stepping over anyone's toes, would it not?

Or maybe lua52 is considered old, and they only support 5.3?

**Comment by jnozsc** [comment]
Thanks all above for understanding this situation. I am totally agree that the perfect solution is `luajit` makes a new release, which seems will not happen soon

@avih  , especially thank for pointing out the option `--lua=`.  I am trying to apply this workaround to homebrew https://github.com/Homebrew/homebrew-core/pull/51335 .  Unfortunately, Homebrew do not like `lua52` https://github.com/Homebrew/homebrew-core/issues/24409 . they have `lua`, which is `lua53` and `lua@5.1`

**Comment by avih** [comment]
> Unfortunately, Homebrew do not like lua52 Homebrew/homebrew-core#24409 . they have lua, which is lua53 and lua@5.1

Then the correct solution is 5.1.

I never tried it so I don't know how it compares to 5.2 in terms of performance. I've been using lua52 on macOS (via macports) for years, and never had an issue (OTOH, in the few times I tried luajit on macOS, I always ended up reverting to lua52 due to $issues).

**Comment by jnozsc** [comment]
Thank all. Close this PR due to the fact that this is an upstream issue,  I commit to `homebrew` to force `homebrew` build with lua51 https://github.com/Homebrew/homebrew-core/pull/51335 , which will make any future release works well

**Pull request closed without merging**